server/src/main/java/org/jboss/as/server/deployment/module/descriptor/DeploymentStructureDescriptorParser.java

```diff
@@ -131,9 +133,9 @@ public void deploy(final DeploymentPhaseContext phaseContext) throws DeploymentU
             //if the parent has already attached parsed data for this sub deployment we need to process it
             if (deploymentRoot.hasAttachment(SUB_DEPLOYMENT_STRUCTURE)) {
                 final ModuleSpecification subModuleSpec = deploymentUnit.getAttachment(Attachments.MODULE_SPECIFICATION);
-                Set<ModuleIdentifier> additionalModules = new HashSet<>();
+                Set<AdditionalModuleSpecification> additionalModules = new HashSet<>();
```

Comment:
``AdditionalModuleSpecification`` does not have equals/hashcode so this is not semantically equivalent to the code that came before.  Can you explain if that is OK and if so, why?

Reply:
By changing `Set<ModuleIdentifier>` to `Set<AdditionalModuleSpecification>`? I think it's fine. https://github.com/wildfly/wildfly-core/pull/2786/files#diff-a84c1903d76fdbbf52a51bfdf6c2fe4f changed `handleDeployment`() method signature to add Set<ModuleIdentifier> additionalModules which is only used in

` if(dependency.getName().startsWith(ServiceModuleLoader.MODULE_PREFIX) && !additionalModules.contains(dependency)) {`

to check if `ModuleIdentifier` dependency is included in `additionalModules`. I update to `AdditionalModuleSpecification` as I need to use the alias list to compare. Method `moduleIncluded()` covers the same check on ModuleIdentifier as before

Reply:
OK, but, this wasn't an opinion question. :)  What makes this OK?  Is it a guarantee that every AdditionalModuleSpecification will have a unique identifier, for example?

Reply:
It seems like it might want to be a list instead of a set.

Reply:
I have updated Set to List.
There is no guarantee at that point right after xml parsing. User can write same identifier for two modules in config file, but if there is either duplicate module-identifier or duplicate module alias e.g. https://gist.github.com/soul2zimate/b2d9bf06f68137c8271db7aaea4c8054, it fails at service installation due to [DuplicateServiceException](https://gist.github.com/soul2zimate/fbcf0cda9b3c4fbd140aaee1e622fbca) 
It is incorrect to contain a duplicate module identifier, Should it fail earlier here?